### PR TITLE
Add OpenGL renderer (squashed)

### LINKED ISSUE #877: Use OpenGL 3.3 for the low-end rendering backend
_edit: see https://github.com/godotengine/godot-proposals/issues/877#issuecomment-633826826 below for updates. Make sure to read both this post and the updates before responding._

**Describe the project you are working on:**
The Godot engine renderer

## The Problem
**Describe the problem or limitation you are having in your project:**

Implementing new 3D features is becoming increasingly difficult as we try to support modern workflows (e.g. HDR, PBR) while maintaining support for OpenGL 2.1/ES 2.0 devices. 

We frequently have to sacrifice quality in our high-end rendering backend in order to support consistent workflows between the high-end (Vulkan) and low-end (OpenGL) backends.

Some common issues that arise:
1. Scenes look very different in GLES3 vs GLES2 because of the use of sRGB vs linear lighting
2. Many optimizations are not possible in GLES2
3. We have to rewrite many parts of core rendering code to support GLES2 devices of varying capabilities
5. Many common post processing and material effects are just not possible in GLES2
6. Common 3D engine features can't work in GLES2, so we don't support them (e.g. terrain, complex particle systems)
7. Our rendering defaults have to be compatible with GLES2 and GLES3, leading to poor quality default scenes overall

### Compatibility with older Devices
Right now the GLES2 backend serves 2 purposes:
1. Compatibility with older devices that only support GLES2
2. Providing a lower-end renderer for lower-end devices and for users who value speed over modern features

When we decided to add the GLES2 backend many devices only supported GLES2. At the time, 40% of Android devices only supported GLES2. Thankfully that is no longer true.

OpenGL 2.1/ ES 2.0 only devices are becoming increasingly rare. On Android over [85% of devices now support GL ES 3.0  or higher](https://developer.android.com/about/dashboards#OpenGL). On iOS, every device supports ES 3.0. On Desktop the overwhelming majority of devices support GL 3.3+ (more than 99% support at least GL 3.3 according to the [Steam hardware survery](https://store.steampowered.com/hwsurvey/Steam-Hardware-Software-Survey-Welcome-to-Steam))

As a result of the rapid adoption of OpenGL/ES 3.x+ the main purpose served by the GLES2 renderer is as a low-end renderer to maximize speed on high and low-end devices. Many users that are choosing to use GLES2 currently are not doing so because their device doesn't support OpenGL 3.3, they are doing so because their device doesn't like the high-end features that are in the Godot GLES3 renderer.

## The Solution
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I propose dropping GLES2 support for 3D. For 2D, we can still use GLES2, but when 3D features are used, we will require GLES3.

Given that the main purpose behind the OpenGL backend will be supporting low-end devices, we will write the GLES3 renderer to target low-end devices. This means it won't have the feature set that that Godot 3.2 GLES3 renderer has, but it should run even faster than the current GLES2 renderer and look much better.

This change will both increase the perceived speed and quality of the low-end renderer, but it will also make development much smoother.

What this allows us to:
1. Standardize workflow between high-end (Vulkan) and low end (OpenGL) backends. This means you can expect a more uniform look when switching between backends.
2. Implement optimizations not present in GLES2 (using instancing, texture arrays, and transform feedback)
3. Have similar lightmap workflow between backends
4. Greatly simplify our engine code which makes maintaining the engine easier
5. Extend more post-processing/materials to the OpenGL backend
6. Support more high-end features that have become standard in 3D engines (e.g. terrain, physical light units, HDR)

The one issue with this proposal is WebGL. WebGL 2.0 is not supported well yet. We will either have to work out a sort of compatibility mode for WebGL, or wait for more widespread adoption of the WebGL 2.0 API.

If we do drop 3D support for GLES2 in 4.0, it will mean that we will have to work harder to support 3.2.x for a long time. Users will still be able to make and export games using Godot 3.2.x and critical bugfixes will still be merged for 3.2.x. This seems to be a fair trade off. If the GLES2 renderer is ported to Godot 4.0, it won't have access to any of the additional features that come with 4.0.

One of the factors we considered when moving to Vulkan was the poor quality of OpenGL ES 3.0 drivers on mobile. At this point, OpenGL ES 3.0 drivers support seems to have improved substantially. Further, basic features are supported very well in the OpenGL ES 3.0 drivers. By limiting ourselves to a limited feature set (closer to GLES2) then we won't bump up against any of the features that are poorly supported by the drivers and we should be able to avoid any of the major problems we had before. 

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
_This section is edited based on comments below_

User will be able to choose between 4 backends:
1. High-end (Vulkan)
2. Low-end/mobile (Vulkan)
3. GLES3 (more limited than current GLES3 backend, similar feature set to current GLES2 backend. Aims for maximum compatibility)
4. GLES2 (2D only to start, may support very basic 3D eventually)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No. 

**Is there a reason why this should be core and not an add-on in the asset library?:**
Yes, this has to do with core engine functionality

### PR DESCRIPTION
**Pull request by Calinou**
Squashed version of https://github.com/godotengine/godot/pull/53150. The original PR is left untouched for future reference.

This closes https://github.com/godotengine/godot-proposals/issues/877.

### CODE REVIEW

**Review by clayjohn — APPROVED** [review]
Provisionally approving. The OpenGL renderer still needs significant cleanup and it is missing most features. But it is in a state where it makes sense to merge into master now and allow people to start working on it in parallel.

**Comment by Calinou** [comment]
I pushed the rename from OpenGL to GLES3. Let me know if it's OK, then I can squash it on top of the most recent commit.

**Comment by reduz** [comment]
@Calinou This looks perfect to me, feel free to squash, very eager to merge it!

**Review by clayjohn — APPROVED** [review]
Reapproval after naming change.  Just a small docs fix remaining to make CI happy.

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks to all involved!

**Comment by reduz** [comment]
This is amazing! Tomorrow will work a bit on the new shader compiler and canvas shader. Will try to port the base RenderingDevice code and get it to work, so @Calinou and others can implement the rest of the features.

File: editor/project_manager.cpp [hunk 1]
```diff
@@ -475,6 +475,13 @@ class ProjectDialog : public ConfirmationDialog {
 					}
 					ProjectSettings::CustomMap initial_settings;
 					initial_settings["rendering/vulkan/rendering/back_end"] = rasterizer_button_group->get_pressed_button()->get_meta(SNAME("driver_name"));
+					if (rasterizer_button_group->get_pressed_button()->get_meta("driver_name") == "Vulkan") {
+						initial_settings["rendering/driver/driver_name"] = "Vulkan";
+					} else {
+						initial_settings["rendering/driver/driver_name"] = "OpenGL3";
+						initial_settings["rendering/textures/vram_compression/import_etc2"] = false;
+						initial_settings["rendering/textures/vram_compression/import_etc"] = true;
+					}
```
Comment (clayjohn) [inline editor/project_manager.cpp:484]: I wonder if this is what is causing #54488 It looks like maybe "driver_name" does not equal "Vulkan" Maybe it should be lowercase?

Comment (Calinou) [reply editor/project_manager.cpp:484]: Indeed, this is likely what's causing the issue. I noticed that we're going for lowercase driver names now, but I forgot to update that part of the code.

Eventually, we may want to do the same for audio drivers.

Comment (clayjohn) [reply editor/project_manager.cpp:484]: See #54488 "driver_name" is actually an int. I recommend we just remove this block altogether until we are ready to expose opengl3 to the project_manager

Comment (Calinou) [reply editor/project_manager.cpp:484]: https://github.com/godotengine/godot/pull/54491

**Pull request merged**